### PR TITLE
SLURM Multi-Node Support

### PR DESCRIPTION
**Pull request by dashstander**
Pairs with [this](https://github.com/EleutherAI/DeeperSpeed/pull/40) DeeperSpeed PR.

The big thing to pay attention to is that when using SLURM and `srun` python's argument parser becomes a lot more picky. The most expedient way to solve this is, when using SLURM, to modify your config files to conform to strict JSON. The regular configs, despite their `yml` extensions, are already pretty close to JSON except for comments and a few capital-T `True`s. There is also an [informative error](https://github.com/EleutherAI/DeeperSpeed/blob/23797d3b704a070d6a9ce5d680e4189e59e0e101/deepspeed/launcher/multinode_runner.py#L143) if someone runs afoul of this limitation.

I think that this is an acceptable solution, though I understand if people would rather not fracture configs based on cluster type in this way. If people have serious reservations I'm open to discussing other options.

There are also small functional changes regarding correctly finding SLURM environment variables.

### CODE REVIEW

**Comment by StellaAthena** [comment]
@Quentin-Anthony or @ShivanshuPurohit can you review this and give it a thumbs up?

File: configs/slurm_local.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "data-path": "data/enron/enron_text_document",
+  "vocab-file": "data/gpt2-vocab.json",
+  "merge-file": "data/gpt2-merges.txt",
+  "save": "checkpoints",
```
Comment (Quentin-Anthony) [inline configs/slurm_local.yml:5]: Add a comment explaining why this is different from `configs/local_setup.yml` (i.e. why you capitalized bools)

Comment (Quentin-Anthony) [reply configs/slurm_local.yml:5]: Oops that's not possible. Comments aren't parsed correctly by srun. We should document this, but elsewhere.

File: megatron/neox_arguments/arguments.py [hunk 2]
```diff
@@ -544,6 +544,12 @@ def configure_distributed_args(self):
             from deepspeed.utils.distributed import mpi_discovery
 
             mpi_discovery()
+        
+        # TODO: Should there just be an argument / attribute like self.deepspeed_mpi for this block?
```
Comment (Quentin-Anthony) [inline megatron/neox_arguments/arguments.py:548]: Add `deepspeed_slurm` to guard this block. Replace the `os.getenv` with it.

**Comment by dashstander** [comment]
Made @Quentin-Anthony 's changes. The pre-commit hooks pass on my local...
Screen Shot 2022-09-21 at 2 59 20 PM

**Review by Quentin-Anthony — APPROVED** [review]

**Pull request merged**